crates/ibc-apps/ics20-transfer/src/context.rs

```diff
@@ -0,0 +1,134 @@
+
+use ibc::core::ics24_host::identifier::{ChannelId, PortId};
+use ibc::prelude::*;
+use ibc::Signer;
+use ibc_app_transfer_types::error::TokenTransferError;
+use ibc_app_transfer_types::{PrefixedCoin, PrefixedDenom, VERSION};
+use sha2::{Digest, Sha256};
+
+/// Methods required in token transfer validation, to be implemented by the host
+pub trait TokenTransferValidationContext {
+    type AccountId: TryFrom<Signer>;
+
+    /// get_port returns the portID for the transfer module.
+    fn get_port(&self) -> Result<PortId, TokenTransferError>;
+
+    /// Returns the escrow account id for a port and channel combination
+    fn get_escrow_account(
+        &self,
+        port_id: &PortId,
+        channel_id: &ChannelId,
+    ) -> Result<Self::AccountId, TokenTransferError>;
+
+    /// Returns Ok() if the host chain supports sending coins.
+    fn can_send_coins(&self) -> Result<(), TokenTransferError>;
+
+    /// Returns Ok() if the host chain supports receiving coins.
+    fn can_receive_coins(&self) -> Result<(), TokenTransferError>;
+
+    /// Validates the sender and receiver accounts and the coin inputs
+    fn send_coins_validate(
+        &self,
+        from_account: &Self::AccountId,
+        to_account: &Self::AccountId,
+        coin: &PrefixedCoin,
+    ) -> Result<(), TokenTransferError>;
+
+    /// Validates the receiver account and the coin input
+    fn mint_coins_validate(
+        &self,
+        account: &Self::AccountId,
+        coin: &PrefixedCoin,
+    ) -> Result<(), TokenTransferError>;
+
+    /// Validates the sender account and the coin input
+    fn burn_coins_validate(
+        &self,
+        account: &Self::AccountId,
+        coin: &PrefixedCoin,
+    ) -> Result<(), TokenTransferError>;
+
+    /// Returns a hash of the prefixed denom.
+    /// Implement only if the host chain supports hashed denominations.
+    fn denom_hash_string(&self, _denom: &PrefixedDenom) -> Option<String> {
+        None
+    }
+}
+
+/// Methods required in token transfer execution, to be implemented by the host
+pub trait TokenTransferExecutionContext: TokenTransferValidationContext {
+    /// This function should enable sending ibc fungible tokens from one account to another
+    fn send_coins_execute(
+        &mut self,
+        from_account: &Self::AccountId,
+        to_account: &Self::AccountId,
+        coin: &PrefixedCoin,
+    ) -> Result<(), TokenTransferError>;
+
+    /// This function to enable minting ibc tokens to a user account
+    fn mint_coins_execute(
+        &mut self,
+        account: &Self::AccountId,
+        coin: &PrefixedCoin,
+    ) -> Result<(), TokenTransferError>;
+
+    /// This function should enable burning of minted tokens in a user account
+    fn burn_coins_execute(
+        &mut self,
+        account: &Self::AccountId,
+        coin: &PrefixedCoin,
+    ) -> Result<(), TokenTransferError>;
+}
+
+// https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-028-public-key-addresses.md
```

Comment:
Do we want to keep this link here? If so, would probably be good to add a bit of context as to why this is here.